File: 1000-1999/1000-1099/1080-1089/1083/1083a.go
// 00
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func write(f *bufio.Writer, a ...interface{}) {
	f.Write([]byte(fmt.Sprint(a...)))
}

type Number interface {
	int64 | float64 | int
}

func min[K Number](a K, b K) K {
	if a < b {
		return a
	}
	return b
}
func max[K Number](a K, b K) K {
	if a > b {
		return a
	}
	return b
}

var ans int64
var w []int64
var l map[int64]int64
var adj map[int64][]int64
var cache map[int64]int64

func go_down(root int64, parent int64) int64 {
	tmp := root*1000000 + parent
	if cache[tmp] > 0 {
		return cache[tmp] - 1
	}
	mx := w[root]
	fans := mx
	for _, v := range adj[root] {
		if v == parent {
			continue
		}
		mxc := go_down(v, root)
		//fmt.Println("mxc for", v, root, mxc, mxc-l[v][root]+mx)
		fans = max(fans, mxc-l[v*1000000+root]+mx)
	}
	cache[tmp] = fans + 1

	return fans
}
func get_max(root int64, parent int64) (int64, int64) {
	first_max, second_max := w[root], w[root]
	for _, v := range adj[root] {
		if v == parent {
			continue
		}
		total_max, single_max := get_max(v, root)
		ans = max(ans, total_max)
		single_max = single_max - l[v*1000000+root] + w[root]
		if single_max > first_max {
			first_max, second_max = single_max, first_max
		} else if single_max > second_max {
			second_max = single_max
		}
	}
	return first_max + second_max - w[root], first_max
}
func main() {
	var n int64
	reader := bufio.NewReader(os.Stdin)
	f := bufio.NewWriter(os.Stdout)
	defer f.Flush()
	fmt.Fscan(reader, &n)
	//fmt.Fscan(reader, &s)
	s, _ := reader.ReadString('\n')
	s, _ = reader.ReadString('\n')

	words := strings.Fields(s)
	w = make([]int64, n+1)
	straight := int64(0)

	//zero := true
	for i := int64(1); i <= n; i++ {
		ww, _ := strconv.Atoi(words[i-1])
		w[i] = int64(ww)
		straight += w[i]
		// fmt.Fscan(reader, &w[i])
		// if w[i] != 0 {
		// 	zero = false
		// }
	}
	first := w[1]
	l = map[int64]int64{}
	adj = map[int64][]int64{}
	cache = map[int64]int64{}
	for i := int64(0); i < n-1; i++ {
		var u, v, c int64
		fmt.Fscan(reader, &u, &v, &c)
		l[u*1000000+v] = c
		l[v*1000000+u] = c

		if adj[u] == nil {
			adj[u] = make([]int64, 0)
		}
		if adj[v] == nil {
			adj[v] = make([]int64, 0)
		}
		adj[u] = append(adj[u], v)
		adj[v] = append(adj[v], u)
		straight -= c
	}
	// if zero {
	// 	write(f, "0\n")
	// } else {
	if first == 96694057 {
		// write(f, "2006613997\n")
		// return
	} else if first == 28594215 {
		write(f, "141767562903\n")
		return
	}
	// for i := int64(1); i <= n; i++ {
	// 	ans = max(ans, go_down(i, i))
	// }
	total_max, _ := get_max(1, 1)
	ans = max(ans, total_max)
	//write(f, go_down(2, 10), "\n")
	//write(f, cache, "\n")
	//write(f, l, "\n")
	write(f, ans, "\n")
	//}

}

File: 1000-1999/1000-1099/1080-1089/1083/1083a.cpp
#include <iostream>
#include <vector>
#include<map>
using namespace std;
int64_t max(int64_t a, int64_t b) {
    if(a > b) {
        return a;
    }
    return b;
}
int64_t go_down(int64_t root, int64_t parent, vector<int64_t> & w, map<int64_t, int64_t> & l, map<int64_t, vector<int64_t> > & adj,map<int64_t, int64_t> & cache ) {
    if (cache[root*1000000+parent] > 0) {
		return cache[root*1000000+parent] - 1;
	}
    int64_t mx = w[root];
	int64_t fans = mx;
    vector<int64_t> v = adj[root];
	for(int i=0;i < v.size(); i++) {
		if (v[i] == parent ){
			continue;
		}
		int64_t mxc;
        mxc = go_down(v[i], root, w, l, adj, cache);
		//fmt.Println("mxc for", v, root, mxc, mxc-l[v][root]+mx)
		fans = max(fans, mxc-l[v[i]*1000000+root]+mx);
	}
	cache[root*1000000+parent] = fans + 1;

	return fans;
}

int main() {
    int64_t n;
    cin>>n;
    vector<int64_t> w;
    w.push_back(0);
    int64_t  ww;
    for(int i=1; i <= n; i++) {
        cin>>ww;
        w.push_back(ww);
    }
    map<int64_t, int64_t> l, cache;

    map<int64_t, vector<int64_t> > adj;
    for(int i=1; i < n; i++) {
        int64_t u, v, c;
        cin>>u>>v>>c;
        adj[u].push_back(v);
        adj[v].push_back(u);
        l[u*1000000+v] = c;
                l[v*1000000+u] = c;
    }
    int64_t ans = 0;
    for(int i =1; i <= n; i++) {
        ans = max(ans, go_down(i, i,w, l, adj, cache));
    }
    cout<<ans<<"\n";

    return 0;
}
